npm/metrics/prometheus-metrics.go

```diff
@@ -53,71 +39,183 @@ const (
 	ipsetInventoryHelp = "The number of entries in each individual IPSet"
 	setNameLabel       = "set_name"
 	setHashLabel       = "set_hash"
+
+	// perf metrics added after v1.4.16
+	// all these metrics have "npm_controller_" prepended to their name
+	operationLabel = "operation"
+	hadErrorLabel  = "had_error"
+
+	policyExecTimeName           = "policy_exec_time"
+	controllerPolicyExecTimeHelp = "Execution time in milliseconds for updating/deleting a network policy. NOTE: for adding, see npm_add_policy_exec_time"
+
+	podExecTimeName           = "pod_exec_time"
+	controllerPodExecTimeHelp = "Execution time in milliseconds for adding/updating/deleting a pod"
+
+	namespaceExecTimeName           = "namespace_exec_time"
+	controllerNamespaceExecTimeHelp = "Execution time in milliseconds for adding/updating/deleting a namespace"
+
+	// TODO add health metrics
+
+	quantileMedian float64 = 0.5
+	deltaMedian    float64 = 0.05
+	quantile90th   float64 = 0.9
+	delta90th      float64 = 0.01
+	quantil99th    float64 = 0.99
+	delta99th      float64 = 0.001
 )
 
+// Gauge metrics have the methods Inc(), Dec(), and Set(float64)
+// Summary metrics have the method Observe(float64)
+// For any Vector metric, you can call With(prometheus.Labels) before the above methods
+//   e.g. SomeGaugeVec.With(prometheus.Labels{label1: val1, label2: val2, ...).Dec()
 var (
-	nodeLevelRegistry    = prometheus.NewRegistry()
-	clusterLevelRegistry = prometheus.NewRegistry()
-	haveInitialized      = false
+	nodeRegistry    = prometheus.NewRegistry()
+	clusterRegistry = prometheus.NewRegistry()
+	haveInitialized = false
+
+	// quantiles e.g. the "0.5 quantile" with delta 0.05 will actually be the phi quantile for some phi in [0.5 - 0.05, 0.5 + 0.05]
+	execTimeQuantiles = map[float64]float64{quantileMedian: deltaMedian, quantile90th: delta90th, quantil99th: delta99th}
+
+	numPolicies          prometheus.Gauge
+	numACLRules          prometheus.Gauge
+	addACLRuleExecTime   prometheus.Summary
+	numIPSets            prometheus.Gauge
+	addIPSetExecTime     prometheus.Summary
+	numIPSetEntries      prometheus.Gauge
+	ipsetInventory       *prometheus.GaugeVec
+	ipsetInventoryLabels = []string{setNameLabel, setHashLabel}
+
+	// controller perf metrics
+	// used to be a regular Summary in v1.4.16 and below
+	addPolicyExecTime       *prometheus.SummaryVec
+	addPolicyExecTimeLabels = []string{hadErrorLabel}
+	// metrics added after v1.4.16
+	controllerPolicyExecTime    *prometheus.SummaryVec
+	controllerPodExecTime       *prometheus.SummaryVec
+	controllerNamespaceExecTime *prometheus.SummaryVec
+	controllerExecTimeLabels    = []string{operationLabel, hadErrorLabel}
+
+	// TODO add health metrics
 )
 
-// InitializeAll creates all the Prometheus Metrics. The metrics will be nil before this method is called.
+type RegistryType bool
+
+const (
+	NodeMetrics    RegistryType = false
```

Comment:
Why are we switching off node metrics by default ?

Reply:
let me switch the type of RegistryType to a string for easier readability